File: ECHA.Web/Controllers/RelatoriosController.fs
namespace ECHA.Web.Controllers

open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Mvc.Rendering
open Microsoft.AspNetCore.Authorization
open System
open System.Threading.Tasks
open EconomiaComHistoria.Core.DTOs
open ECHA.Web.Services
open EconomiaComHistoria.Core.Enums

[<Authorize(Roles = "Admin,Editor,SuperAdmin")>]
type RelatoriosController (apiClient: ApiClient) =
    inherit Controller()

    member private this.GetToken() =
        let claim = this.User.FindFirst("AccessToken")
        if isNull claim then null else claim.Value

    // Carrega dropdowns de escolas e turmas (opcionalmente filtrando por escola)
    member private this.CarregarDropdownsAsync (token: string, ?escolaId: int) =
        task {
            let! escolas = apiClient.ListEscolasAsync(token)
            this.ViewData.["Escolas"] <- SelectList(escolas, "Id", "Nome")

            let! turmas = apiClient.ListTurmasAsync(token, ?escolaId = escolaId)
            this.ViewData.["Turmas"] <- SelectList(turmas, "Id", "Nome")
        }

    [<HttpGet>]
    member this.Index (escolaId: Nullable<int>) =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let escolaIdOpt = if escolaId.HasValue then Some escolaId.Value else None
                let! lista = apiClient.ListarRelatoriosAsync(t, ?escolaId = escolaIdOpt)
                let listaDotNet = new System.Collections.Generic.List<RelatorioListaDto>(lista)
                return this.View(listaDotNet) :> IActionResult
        }

    [<HttpGet>]
    member this.Criar (escolaId: Nullable<int>) =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let escolaIdOpt = if escolaId.HasValue then Some escolaId.Value else None

                do! this.CarregarDropdownsAsync(t, ?escolaId = escolaIdOpt)

                let dto = new SolicitarRelatorioDto("", "CSV", escolaId, Nullable<int>(), Nullable<DateTime>(), Nullable<DateTime>())
                return this.View(dto) :> IActionResult
        }

    [<HttpPost>]
    member this.Criar (request: SolicitarRelatorioDto) : Task<IActionResult> =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let escolaIdOpt = if request.EscolaId.HasValue then Some request.EscolaId.Value else None

                if String.IsNullOrWhiteSpace(request.Titulo) then
                    this.TempData["ErrorMessage"] <- "Título é obrigatório."
                    do! this.CarregarDropdownsAsync(t, ?escolaId = escolaIdOpt)
                    return this.View(request) :> IActionResult
                else
                    let! result = apiClient.SolicitarRelatorioAsync(request, t)
                    match result with
                    | Some status ->
                        this.TempData["SuccessMessage"] <- "Relatório solicitado com sucesso! A geração pode demorar alguns minutos."
                        // ✅ CORREÇÃO: usar objeto anónimo para redirecionamento
                        return this.RedirectToAction("Status", {| id = status.Id |}) :> IActionResult
                    | None ->
                        this.TempData["ErrorMessage"] <- "Erro ao solicitar relatório."
                        do! this.CarregarDropdownsAsync(t, ?escolaId = escolaIdOpt)
                        return this.View(request) :> IActionResult
        }

    [<HttpGet>]
    member this.Status (id: int) =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                let! status = apiClient.GetRelatorioStatusAsync(id, t)
                match status with
                | Some s -> return this.View(s) :> IActionResult
                | None -> return this.NotFound() :> IActionResult
        }

    [<HttpGet>]
    member this.Download (id: int) =
        task {
            let token = this.GetToken()
            match token with
            | null -> return this.Unauthorized() :> IActionResult
            | t ->
                // 1. Procura o estado do relatório para saber o formato (PDF/CSV) e se está concluído
                let! statusOpt = apiClient.GetRelatorioStatusAsync(id, t)
                match statusOpt with
                | None -> return this.NotFound() :> IActionResult
                | Some status ->
                    if status.Estado <> EstadoRelatorio.Concluido then
                        this.TempData["ErrorMessage"] <- "O relatório ainda não está pronto para download."
                        return this.RedirectToAction("Status", {| id = id |}) :> IActionResult
                    else
                        // 2. Procura os bytes do ficheiro através do ApiClient
                        // Nota: Certifique-se de que o seu ApiClient tem o método DownloadRelatorioAsync implementado
                        let! dadosOpt = apiClient.DownloadRelatorioAsync(id, t)
                        match dadosOpt with
                        | None -> 
                            this.TempData["ErrorMessage"] <- "Ficheiro não encontrado no servidor de API."
                            return this.RedirectToAction("Status", {| id = id |}) :> IActionResult
                        | Some (bytes: byte[]) ->
                            let extensao = if status.Tipo.ToLower() = "csv" then "csv" else "pdf"
                            let contentType = if extensao = "csv" then "text/csv" else "application/pdf"
                            let nomeFicheiro = sprintf "relatorio_%d.%s" id extensao
                            
                            return this.File(bytes, contentType, nomeFicheiro) :> IActionResult
        }

File: ECHA.API/Controllers/RelatoriosController.cs
using EconomiaComHistoria.Core.DTOs;
using EconomiaComHistoria.Core.Enums;
using EconomiaComHistoria.Core.Interfaces;
using EconomiaComHistoria.Infrastructure.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace ECHA.API.Controllers;

[ApiController]
[Route("api/relatorios")]
[Authorize(Roles = "Admin,Editor,SuperAdmin")]
public class RelatoriosController : ControllerBase
{
    private readonly IRelatorioService _relatorioService;
    private readonly IAuditoriaService _auditoriaService;

    public RelatoriosController(IRelatorioService relatorioService, IAuditoriaService auditoriaService)
    {
        _relatorioService = relatorioService;
        _auditoriaService = auditoriaService;
    }

    private bool TryGetUserId(out int userId)
    {
        var value = User.FindFirst(ClaimTypes.NameIdentifier)?.Value
                 ?? User.FindFirst("sub")?.Value;
        return int.TryParse(value, out userId);
    }

    [HttpGet]
    public async Task<ActionResult<List<RelatorioListaDto>>> ListarRelatorios(
        [FromQuery] int? escolaId,
        CancellationToken ct)
    {
        if (!TryGetUserId(out var userId))
            return Unauthorized();

        var lista = await _relatorioService.ListarRelatoriosAsync(userId, escolaId, ct);
        return Ok(lista);
    }

    [HttpPost("gerar")]
    public async Task<ActionResult<RelatorioStatusDto>> GerarRelatorio(
        [FromBody] SolicitarRelatorioDto request,
        CancellationToken ct)
    {
        if (!TryGetUserId(out var userId))
            return Unauthorized();

        if (string.IsNullOrWhiteSpace(request.Titulo))
            return BadRequest(new { message = "Título é obrigatório." });

        // Validar que a data de fim não é anterior à data de início
        if (request.Fim.HasValue && request.Inicio.HasValue && request.Fim.Value < request.Inicio.Value)
            return BadRequest(new { message = "A data de fim não pode ser anterior à data de início." });

        // Validar que a data de fim não seja anterior à data de início
        if (request.Fim.HasValue && request.Inicio.HasValue && request.Fim.Value < request.Inicio.Value)
            return BadRequest(new { message = "A data de fim não pode ser anterior à data de início." });

        var status = await _relatorioService.SolicitarRelatorioAsync(userId, request, ct);
        await _auditoriaService.RegistarAsync(
            userId,
            "SolicitarRelatorio",
            "Relatorio",
            status.Id,
            null,
            $"Título: {request.Titulo}, Tipo: {request.Tipo}",
            HttpContext
        );
        return CreatedAtAction(nameof(GetStatus), new { id = status.Id }, status);
    }

    [HttpGet("{id}/status")]
    public async Task<ActionResult<RelatorioStatusDto>> GetStatus(int id, CancellationToken ct)
    {
        if (!TryGetUserId(out var userId))
            return Unauthorized();

        var status = await _relatorioService.GetStatusAsync(id, ct);
        if (status == null)
            return NotFound();

        return Ok(status);
    }

    [HttpGet("{id}/download")]
    public async Task<IActionResult> Download(int id, CancellationToken ct)
    {
        if (!TryGetUserId(out var userId))
            return Unauthorized();

        var status = await _relatorioService.GetStatusAsync(id, ct);
        if (status == null || status.Estado != EstadoRelatorio.Concluido)
            return NotFound(new { message = "Relatório não encontrado ou não está pronto para download." });

        var dados = await _relatorioService.DownloadRelatorioAsync(id, ct);
        if (dados == null)
            return NotFound(new { message = "Ficheiro não encontrado." });

        // Sempre retornar como CSV agora que geramos apenas CSV
        var nomeFicheiro = $"relatorio_{id}.csv";

        await _auditoriaService.RegistarAsync(
            userId,
            "DescarregarRelatorio",
            "Relatorio",
            id,
            null,
            null,
            HttpContext
        );

        return File(dados, "text/csv", nomeFicheiro);
    }
}